presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java

```diff
@@ -399,6 +772,107 @@ public void partitionPage(Page page)
             flush(false);
         }
 
+        public void ariaPartitionPage(Page page)
+        {
+            populatePositionsForEachPartition(page);
+
+            decodePage(page);
+
+            for (int i = 0; i < partitionData.length; i++) {
+                PartitionData target = partitionData[i];
+                target.appendRows(decodedBlocks, containsNonFixedWidthColumns, rowSizeForfixedWidthBlocks, variableWidthBlockChannels, outputBuffer);
+            }
+        }
+
+        private void calculateRowSizeForFixedWidthBlocks()
+        {
+            rowSizeForfixedWidthBlocks = 0;
+            containsNonFixedWidthColumns = false;
+            for (int i = 0; i < sourceTypes.size(); i++) {
+                Type type = sourceTypes.get(i);
+                Integer serializedBytesPerRowForCurrentBlock = FIXED_WIDTH_TYPE_SERIALIZED_BYTES.getOrDefault(type.getClass(), 0);
+                rowSizeForfixedWidthBlocks += serializedBytesPerRowForCurrentBlock;
+
+                if (serializedBytesPerRowForCurrentBlock == 0) {
+                    containsNonFixedWidthColumns = true;
+                    variableWidthBlockChannels.add(i);
+                }
+            }
+        }
+
+        private void decodePage(Page page)
+        {
+            int channelCount = page.getChannelCount();
+
+            if (decodedBlocks == null) {
+                decodedBlocks = new DecodedObjectNode[channelCount];
+            }
+
+            for (int i = 0; i < page.getChannelCount(); i++) {
+                decodedBlocks[i] = decodeBlock(page.getBlock(i));
+            }
+        }
+
+        // Dict/RLE is a separate node. There can't be consequetive Dict/RLE nodes because we call flatten
+        private DecodedObjectNode decodeBlock(Block block)
+        {
+            Block decodedBlock = flattener.flatten(block).get();
```

Comment:
I think you need to use try-with-resources to make sure the BlockLease returned from flattener.flatten(block) gets closed. CC: @tdcmeehan 

```
try (BlockLease lease = flattener.flatten(block)) {
                Block decodedBlock = lease.get();
```